libs/server/AOF/ReplayCoordinator/AofReplayCoordinator.cs

```diff
@@ -0,0 +1,283 @@
+            /// <param name="asReplica"></param>
+            internal void ProcessFuzzyRegionOperations(long storeVersion, bool asReplica)
+            {
+                var fuzzyRegionOps = aofReplayContext.fuzzyRegionOps;
+                if (fuzzyRegionOps.Count > 0)
+                    logger?.LogInformation("Replaying {fuzzyRegionBufferCount} records from fuzzy region for checkpoint {newVersion}", fuzzyRegionOps.Count, storeVersion);
+                foreach (var entry in fuzzyRegionOps)
+                {
+                    fixed (byte* entryPtr = entry)
+                        _ = aofProcessor.ReplayOp(aofProcessor.basicContext, aofProcessor.objectStoreBasicContext, entryPtr, entry.Length, asReplica);
+                }
+            }
+
+            /// <summary>
+            /// Process fuzzy region transaction groups
+            /// </summary>
+            /// <param name="ptr"></param>
+            /// <param name="asReplica"></param>
+            internal void ProcessFuzzyRegionTransactionGroup(byte* ptr, bool asReplica)
+            {
+                Debug.Assert(aofReplayContext.txnGroupBuffer != null);
+                // Process transaction groups in FIFO order
+                if (aofReplayContext.txnGroupBuffer.Count > 0)
+                {
+                    var txnGroup = aofReplayContext.txnGroupBuffer.Dequeue();
+                    ProcessTransactionGroup(ptr, asReplica, txnGroup);
+                }
+            }
+
+            /// <summary>
+            /// Process provided transaction group
+            /// </summary>
+            /// <param name="asReplica"></param>
+            /// <param name="txnGroup"></param>
+            internal void ProcessTransactionGroup(byte* ptr, bool asReplica, TransactionGroup txnGroup)
+            {
+                if (!asReplica)
+                {
+                    // If recovering reads will not expose partial transactions so we can replay without locking.
+                    // Also we don't have to synchronize replay of sublogs because write ordering has been established at the time of enqueue.
+                    ProcessTransactionGroupOperations(aofProcessor, aofProcessor.basicContext, aofProcessor.objectStoreBasicContext, txnGroup, asReplica);
+                }
+                else
+                {
+                    var txnManager = aofProcessor.respServerSession.txnManager;
+
+                    // Start by saving transaction keys for locking
+                    SaveTransactionGroupKeysToLock(txnManager, txnGroup);
+
+                    // Start transaction
+                    _ = txnManager.Run(internal_txn: true);
+
+                    // Process in parallel transaction group
+                    ProcessTransactionGroupOperations(aofProcessor, txnManager.LockableContext, txnManager.ObjectStoreLockableContext, txnGroup, asReplica);
+
+                    // Commit (NOTE: need to ensure that we do not write to log here)
```

Comment:
Can we add a debug assert or similar here to make sure we do not break this invariant of not writing to log here.